Sweetandsour2

bthaeler opened the pull request:
This is a new hook system that replaces the embedded perf code. This will be the foundation for a new debugger interface.

bthaeler committed:
build hook system into mojito
remove perf markers, and replace with hooks.

bthaeler committed:
build hook system into mojito
remove perf markers, and replace with hooks.

bthaeler committed:
changed to null context for clients.
updated documents.

bthaeler committed:
more stuff.

bthaeler committed:
Merge branch 'develop' of git://github.com/yahoo/mojito into sweetandsour2

Conflicts:
	lib/app/autoload/store.server.js

bthaeler committed:
add hook to template renderiing.

caridy commented:
@bthaeler a couple of questions:
- did you validate the unit tests? I remember when I put in place the perf there were few issues with the tests due to the fact that every script is using Y.mojito.hooks, which is undefined if it is not part of the testsuite mocks.
- is the current output format of the perf still preserved? if not, we need to verify with @drewfish about this impact of that since we use it in the current profiler and potentially in our CI to monitor performance, but I'm not sure about that last part.

caridy commented:
According to @drewfish we are not using perf and profiler at all in our CI, so, feel free to break it, and even remove it in favor of the new profiler.

So, the only thing pending is the tests to pass.

bthaeler commented:
The existing perf markers are re-implemented using the hook system. So they should continue to work just fine.

It wasn't very clear to me how to run the unit tests. So I didn't. Looking at the Travis build output on github, it looks like some of the tests do complain. I can try and fix those. Is there a good current document explaining how to setup a test system for running the unit tests?

caridy commented:
Few pointers:
- everyone who uses hooks should requires mojito-hooks.
- mock it in https://github.com/yahoo/mojito/blob/develop/tests/base/mojito-test.js to do nothing when running tests.
- to run tests, check:
  cd tests/ 
  ./run.js help
- more info about tests here: https://github.com/yahoo/mojito/tree/develop/tests

bthaeler committed:
added some missing hook dependencys.
added stub to test harnas

bthaeler commented:
I added some missing mojito-hooks dependency to places, and I added a stuff for the hook system to the test base. I think the unit tests are fine now.

caridy commented on tests/base/mojito-test.js:
```diff
@@ -26,6 +26,12 @@ YUI.add('mojito', function(Y, NAME) {
     Y.namespace('mojito.addons.viewEngines');
+    // HookSystem::StartBlock
+    Y.mojito.hooks = {
+        hook: function () {},
```
I think `enableHookGroup` is also missing from this. Even though only perf is calling it conditionally, I think we will have more scripts calling it, in which case we need a mock for testing.

caridy replied:
Any update on this one?

bthaeler replied:
yes, I added a stub function.

```
    enableHookGroup: function () {}
```

caridy commented on lib/mojito.js:
```diff
@@ -281,6 +281,13 @@ MojitoServer.prototype._configureAppInstance = function(app, options) {
         });
+        // HookSystem::StartBlock
+        if (!req.hook) {
+            req.hook = null;
```
What is the purpose of setting it to `null` when it is already a falsy value?

bthaeler replied:
From my, and your performance tests, it is more efficient to test if a field is null vs undefined. IE 'if (x.y)' is more efficient if x.y = null vs x.y is just undefined.

I am just trying to make the hook system as low impact as possible.

caridy replied:
ok, we can revisit this later on after the first merge.

caridy commented on lib/mojito.js:
```diff
@@ -281,6 +281,13 @@ MojitoServer.prototype._configureAppInstance = function(app, options) {
+            req.hook = null;
+        }
+        outputHandler.hook = req.hook;
+        command.instance.hook = req.hook;
```
Any particular reason for adding hook into the `instance` instead of `command`. In theory, an instance's action can only be executed if there is a command to be dispatched. In my mind, `hook` is more tied to `command` than it is to `instance`. In any case, you should have access to `command` from everywhere during the dispatch process, just like instance. In the server, this is just a technicallity, but in the client, that instance is persistent, and command is created every time an action is invoked.

caridy commented on lib/app/autoload/mojito-client.client.js:
```diff
@@ -629,6 +642,12 @@ YUI.add('mojito-client', function(Y, NAME) {
 
+            // HookSystem::StartBlock
+            if (Y.mojito.hooks) {
+                outputHandler.hook = Y.mojito.hooks.client_ctx;
```
`client_ctx` sounds weird here, specially when you assign it to `outputHandler.hook`. Is client_ctx  the persistent hook in the client side? If yes, can we rename it? We don't want to keep overloading the `context` term in mojito, we already use it for many things.

bthaeler replied:
Originally this was for the clients. But this also works for people who want to enable a hook global, without associating this with a request. So how about 'global_hooks'.

Each callback will have its own unique this state per tool. This is stored in this object. So do we want 'global_hooks' or 'global_hooks_ctx'?

caridy commented on lib/app/autoload/hooks.common.js:
```diff
@@ -0,0 +1,161 @@
+        }
+        map_hook_tool[hook][tool] = cb;
+
+        Y.mojito.hooks.hook = Y.mojito.hooks.real_hook;
```
Why is this happening here? This sounds like a leak-prompt structure, because any request can potentially change this. If the hook is something created per request, then it should be an inmutable object.

bthaeler replied:
Not sure what your concern is here. Hooks are statically defined. Groups of hooks are enabled on a per request basis. What hooks people are interested in is static.

caridy replied:
My question is why hook gets real_hook whenever a new hook is register? is this something to have "hook" defined only if there is at least one hook registered? I'm just curious about this statement.

bthaeler replied:
Yes, the idea is that if you are not using any code that needs a hook, the hook system defaults to an empty function.

caridy replied:
ok

caridy commented on lib/app/autoload/hooks.common.js:
```diff
@@ -0,0 +1,161 @@
+        var ret = {};
+        // Y.log("enable tool: " + tool);
+
+        // if called in client, use single global context.
```
In my experience, trying to accomodate a component to be `common` when in reality the behavior of the client and server are very different will bring problems. I wonder if we should split this into server and client.

bthaeler replied:
If I rename this to 'global_hooks' I think this resolved this.

caridy commented on lib/app/autoload/hooks.common.js:
```diff
@@ -0,0 +1,161 @@
+    };
+
+    Y.mojito.hooks.real_hook = function (hook, ctx) {
+        // Y.log("in hook handler: " + hook);
```
Please, preserve all Y.log(msg, 'info or debug or error', NAME), instead of commenting them.

caridy commented on lib/app/autoload/hooks.common.js:
```diff
@@ -0,0 +1,161 @@
+ * List of hooks
+ *
+ * <ul>
+ * <li>Y.mojito.hooks.hook('adapterBuffer', hook_ctx, 'start', this);
```
Why do we have these docs here? this is the hook utility, it will be difficult to keep this in sync with all the hooks defined across the mojito or the app. We can create a tool to generate such as list, or we can document them in a form of YUI events on site, rather than here.

caridy replied:
any update on this one?

bthaeler replied:
Sounds like you are proposing two options. One is to create a script that runs at build time to generate this list, and the second one is using some kind of YUI event to do this? I don't now what would be better. Is this something we want to solve now or latter? If you feel strongly about this, I can try and figure something out.

caridy commented on lib/app/addons/view-engines/hb.server.js:
```diff
@@ -39,26 +39,32 @@ YUI.add('mojito-hb', function(Y, NAME) {
-                    'time to render a template', tmpl),
-                handler = function (err, obj) {
-                    var output;
+                handler;
```
Any particular reason for moving handler's function definition down rather than in the var statement?

bthaeler replied:
No, i will fix.

caridy commented on lib/app/addons/ac/composite.common.js:
```diff
@@ -204,8 +209,10 @@ callback({
-                meta = {},
-                perf;
+                // HookSystem::StartBlock
+                self = this,
```
my convention, we use `my = this` instead of `self = this`. Don't ask my why :)

And by the way, don't need to define the comments for that line, it is ok to have `my` defined even when u don't use it later on.

caridy commented on lib/app/addons/ac/composite.common.js:
```diff
@@ -14,19 +14,24 @@
         this.__meta = [];
-        this.__perf = Y.mojito.perf.timeline(NAME, 'child', 'the whole child', id);
+        // HookSystem::StartBlock
+        this.__hook_ctx = hook_ctx;
```
Can you elaborate more about this `__hook_ctx` very private reference?

caridy commented on lib/app/addons/ac/composite.common.js:
```diff
@@ -14,19 +14,24 @@
 YUI.add('mojito-composite-addon', function(Y, NAME) {
 
-    function AdapterBuffer(buffer, id, callback) {
+    function AdapterBuffer(buffer, id, callback, hook_ctx) {
```
This is the only part that I really don't like. Extending the signature of the class to add the hook reference after the callback is not pretty. Let's try to find another way to keep that reference around.

caridy commented on lib/app/addons/ac/composite.common.js:
```diff
@@ -335,7 +346,7 @@ callback({
 
                     childAdapter = new AdapterBuffer(buffer, childName,
-                            callback);
+                            callback, this.adapter.hook);
```
same as above, we should not do this here. Let's find another way for the adapter to pull a reference of the current hook.

caridy commented:
@bthaeler that's all for now. Let's work together to accomodate those details so we can finally merge it.

bthaeler committed:
some changes recommended by Caridy
Plus fixes for end hook points

bthaeler committed:
added enableHookGroup to test stubs

bthaeler committed:
cliend needs to use global_hooks_ctx too.

caridy commented:
+1

@bthaeler I think we are ready to merge it. Since we have a release coming up on monday, I will hold this until after that. 

/cc @jenny

bthaeler commented:
Any idea when you will do the merge for this pull request?

caridy commented:
today or tomorrow! 0.5.2 is out!
